src/inference/src/cpp_interfaces/interface/ie_iplugin_internal.cpp

```diff
@@ -340,75 +342,96 @@ std::unordered_set<std::string> GetSupportedNodes(
+    for (auto&& name : unsupported) {
+        supported.erase(name);
     }
-    for (auto&& node : model->get_ops()) {
-        if (InferenceEngine::details::contains(supported, node->get_friendly_name())) {
-            for (auto&& inputNodeOutput : node->input_values()) {
-                if (ov::op::util::is_constant(inputNodeOutput.get_node()) ||
-                    ov::op::util::is_parameter(inputNodeOutput.get_node())) {
-                    supported.emplace(inputNodeOutput.get_node()->get_friendly_name());
+    // Walk over transformed model for special handing of Parameters/Constants/Results
+    for (auto&& op : ops) {
+        // Mark Parameter/Constants/Result as supported if they are connected
+        // to supported node
+        if (InferenceEngine::details::contains(supported, op->get_friendly_name())) {
+            for (auto&& input : op->input_values()) {
+                if (ov::op::util::is_constant(input.get_node()) || ov::op::util::is_parameter(input.get_node())) {
+                    auto names = get_names_set(input.get_node_shared_ptr());
+                    supported.insert(names.begin(), names.end());
                 }
             }
-            for (auto&& outputs : node->outputs()) {
-                for (auto&& outputNodeInput : outputs.get_target_inputs()) {
-                    if (ov::op::util::is_output(outputNodeInput.get_node())) {
-                        supported.emplace(outputNodeInput.get_node()->get_friendly_name());
+            for (auto&& outputs : op->outputs()) {
+                for (auto&& input : outputs.get_target_inputs()) {
+                    if (ov::op::util::is_output(input.get_node())) {
+                        supported.emplace(input.get_node()->get_friendly_name());
                     }
                 }
             }
         }
-
-        if (ov::op::util::is_constant(node) || ov::op::util::is_parameter(node)) {
-            if (node->output(0).get_target_inputs().size() &&
-                !InferenceEngine::details::contains(
-                    supported,
-                    node->output(0).get_target_inputs().begin()->get_node()->get_friendly_name())) {
-                supported.erase(node->get_friendly_name());
+        // Mark Parameter/Constants/Result as supported if they are have no
+        // supported consumers/sources
+        if (ov::op::util::is_constant(op) || ov::op::util::is_parameter(op)) {
+            bool all_consumers_unsupported = true;
+            for (auto&& input : op->output(0).get_target_inputs()) {
+                auto check_node = input.get_node()->shared_from_this();
+                for (auto& name : get_names_set(check_node)) {
+                    if (InferenceEngine::details::contains(supported, name)) {
+                        all_consumers_unsupported = false;
+                        break;
+                    }
+                }
+            }
+            if (all_consumers_unsupported) {
```

Comment:
Can this logic calls non-optimal graph splitting in the Hetero plugin?

Reply:
> Can this logic calls non-optimal graph splitting in the Hetero plugin?

Vice versa: this logic is trying to avoid having having Constant on target device. Previously we had such logic, but it used **original** model instead **transformed**. I described it in PR: if, for example, we have constant subgraph, which is folded, all fused ops are marked as supported, it means that Constant from original graph will have supported consumers (fused). But actual folded Constant can have unsupported ones.